### PR TITLE
Credentials support

### PR DESCRIPTION
**Pull request by DerNamenlose**
This PR implements the support for credentials when interacting with private registries.

### CODE REVIEW

File: crates/oro-client/src/credentials.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        f.write_fmt(format_args!("username: {}, password: ***", self.username))
+    }
+}
+
+/**
+ * Different credential types supported by orogene
+ */
+#[derive(Clone, Debug)]
+pub enum Credentials {
+    Basic(BasicAuthCredentials),
```
Comment (zkat) [inline crates/oro-client/src/credentials.rs:25]: why not just `Basic { username: String, password: String }`?

File: crates/oro-client/src/credentials.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    Token(String),
+}
+
+impl TryFrom<&HashMap<&String, &String>> for Credentials {
+    type Error = OroClientError;
+
+    fn try_from(value: &HashMap<&String, &String>) -> Result<Self, Self::Error> {
+        // TODO I don't like those &...to_string() constructs -> find a better type for the HashMap!
+        if let Some(&token) = value.get(&"token".to_string()) {
```
Comment (zkat) [inline crates/oro-client/src/credentials.rs:34]: doesn't `value.get("token")` work as-is?...

File: crates/oro-client/src/credentials.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            Ok(Self::Token(token.to_owned()))
+        } else if value.contains_key(&"username".to_owned())
```
Comment (zkat) [inline crates/oro-client/src/credentials.rs:36]: `else if let (Some(username), Some(password)) = (value.get("username", value.get("password")) { ... }`

... is probably a nicer way to express this :)

File: src/apply_args.rs [hunk 4]
```diff
@@ -99,6 +99,9 @@ pub struct ApplyArgs {
     #[arg(from_global)]
     pub registry: Url,
 
+    #[arg(from_global)]
+    pub credentials: Vec<(String, String, String)>,
```
Comment (zkat) [inline src/apply_args.rs:103]: You'll want to actually pass these in through the `configured_maintainer` function below.

File: src/nassun_args.rs [hunk 5]
```diff
@@ -15,6 +15,9 @@ pub struct NassunArgs {
     #[arg(from_global)]
     registry: Url,
 
+    #[arg(from_global)]
+    credentials: Vec<(String, String, String)>,
```
Comment (zkat) [inline src/nassun_args.rs:19]: you'll want to actually apply these in `to_nassun` below.

File: crates/nassun/src/client.rs [hunk 6]
```diff
@@ -57,6 +58,15 @@ impl NassunOpts {
...
     }
 
+    /// Set the credential-config for this instance. The config has the form (registry,key,value)
+    /// where registry is the host of the registry, key is the key of a single entry in the respective
+    /// config JSON object and value is the corresponding value. A single registry may have multiple entries
+    /// in this vector (e.g. "username" & "password"). The format will be handled by by OroClient.
+    pub fn credentials(mut self, credentials: Vec<(String, String, String)>) -> Self {
+        self.credentials = credentials;
+        self
+    }
```
Comment (zkat) [inline crates/nassun/src/client.rs:68]: you want this option in node-maintainer, too, and make sure it gets passed down directly to nassun_opts.

File: crates/oro-client/src/credentials.rs (hunk 2 above)
Comment (DerNamenlose) [reply crates/oro-client/src/credentials.rs:34]: Not for `HashMap<&String, &String>`. For `HashMap<String, String>`, however... I've changed the type of the map.

File: crates/oro-client/src/credentials.rs (hunk 1 above)
Comment (DerNamenlose) [reply crates/oro-client/src/credentials.rs:25]: Because I forgot that this was possible. Need sleep or something.

File: src/apply_args.rs (hunk 4 above)
Comment (DerNamenlose) [reply src/apply_args.rs:103]: I'll just nod and do for now. To be honest: I haven't full understood yet how the whole flow with the arguments is. And the little I did understand is basically gone, now that I haven't looked at the code in 2 weeks.

File: src/nassun_args.rs (hunk 5 above)
Comment (DerNamenlose) [reply src/nassun_args.rs:19]: Done.

File: crates/nassun/src/client.rs (hunk 6 above)
Comment (DerNamenlose) [reply crates/nassun/src/client.rs:68]: I've added a `fn credentials(...)` to `NodeMaintainerOptions` and just hand those over to the contained `nassun_opts` there.

**Comment by zkat** [comment]
heya! It's been a minute since I've been spending time on orogene, and I'm just going through old PRs and seeing what's up. How's this going?

**Comment by DerNamenlose** [comment]
Hey. Sorry, I don't see me getting around doing much here in the foreseeable future. I'm happy to let this lie, if you are, but I'm also equally happy for someone else to take this over or close it because it's outdated or something.

**Comment by zkat** [comment]
Sounds good! Thanks for letting me know. I might pick up where you left off when I find some time and take it over the finish line :)

**Comment by zkat** [comment]
Closing in favor of https://github.com/orogene/orogene/pull/292.

Thanks for getting this started. I've kept your (credited) commit in that other PR, so you'll still get credit for the work you did!

**Pull request closed without merging**